Need feedback on database structure

jillianw opened the pull request:
Here's my Google sheet:

https://docs.google.com/spreadsheets/d/1Vw75Hh-x6ReEEFHmVNJ3jrboVk5YQipN4cMyIDmbPZU/edit#gid=0

Before I tab everything out correctly and complete the tables for all competitions, could you let me know if I have any major design flaws? I feel like I'm covering all the bases, but I want to be sure.

Thanks

halfghaninne commented:
Some things to consider:
- Information is being repeated in your competitors table. Add separate college and conferences table and refer elsewhere to their specific rows by id.
- How does the awards table access the ids of the top-ranking students in each event? I might expect in this table to see columns like `event_id` `competitor_id` `rank`. I might also expect the information under `name` in the awards table to be included in some events table, referenced by `event_id`

jillianw commented:
Is the last Rank column on the Event Results table necessary? I feel like I would use code to determine rank based on Time instead.

sumeetjain commented:
- In the competitors table, no need to indicate conference_id. The competitor belongs to a college, and the college belongs to a conference. So the competitor belongs to a conference "through" their school.

> Is the last Rank column on the Event Results table necessary? I feel like I would use code to determine rank based on Time instead.

It could go either way–you're absolutely right. I suggest leaving it off.

sumeetjain commented:
Once you make these changes, no need to submit for approval. You are :+1: Tables are excellent.

jillianw commented:
Just wondering if I'm on the right track with this. Already got the schema set up, added ActiveRecord to table classes, and I've added some forms in the various view files and started trying to work on the home, conferences and colleges controllers. What I'm trying to do is make the user enter form data on one page after another. The home page has a link to register a competitor, and from there it flows like this:

conferences --> colleges --> competitors --> events --> confirmation

Can you tell me if I'm doing this right or if this setup is a bad idea? I feel like the html I wrote should go from page to page as intended, but I'm less sure about the correctness of the controllers.

Thanks

sumeetjain commented:
For now, stick with a simpler (though more circuitous) approach. Treat the creation of each kind of record as a separate interaction. So the admin would create a competitor using a form, and when it's submitted, they're taken back to some "hub" page where they can create something else. Like a conference. Or another competitor. Or a signup, etc.

The idea is that each record's creation form should not depend on the just-submitted previous form.

It's also worth considering that the admin will probably begin by creating the existing conferences first. Otherwise, when they try to indicate the conference a team belongs to, they won't have any options to pick from.

Also, consider this scenario: The admin needs to add a competitor from a school, but that school does not yet exist. The way I'm suggesting this would happen is:
1. The admin adds the school.
2. The admin adds the competitor.

Let me know if this needs clarification! I haven't had coffee yet.

jillianw commented:
I've got controllers set up for home/add_controller/added_controller for adding events to the database. Can you let me know if this looks alright before I proceed with the rest of the controllers?

Thanks!

sumeetjain commented:
Controllers look good! Are you doing all of the controllers before views? I noticed that the controller action for adding a new event _does_ have a view that it references, but that [view](https://github.com/jillianw/web-app__swimming-competition/blob/master/app/views/add/add_events.erb) is empty.

If you are doing all controllers before views, that's fine. Just want to make sure I'm on the same page as you.

jillianw commented:
I'm stuck here. I can't figure out how to write my controller actions that will process forms for adding awards/events/competitors/etc. to the database. I know that the stuff inside the params brackets is incorrect. I didn't understand params when it came up the first time and I still don't. I've tried watching back through the videos but I can't find what I'm looking for that is applicable to what I'm doing. I'm spinning my wheels and wasting a lot of time. Could use some direction. I plan on spending the long weekend getting caught back up on all the stuff I should know already.

Thanks

sumeetjain commented:
On my way!

sumeetjain commented:
The only thing I can't find in your code is where you provide the event/competitor IDs when creating new `EventDetail` or `Award` records.

I see the form fields, e.g. in **app/views/add/add_awards.erb**:

``` erb
    Select the event id:
    <input type="number" name="event_id"> 

    Select the competitor's id:
    <input type="number" name="competitor_id">

    Select the competitor's rank:
    <input type="number" name="rank">
```

But the controller action which that form submits to only uses the _rank_:

``` ruby
MyApp.get "/award_added" do 
  x = Award.new
  x.rank = params["rank"]
  x.save

  erb :"added/award_added"
end
```

I would have expected that to be:

``` ruby
MyApp.get "/award_added" do 
  x = Award.new
  x. competitor_id = params["competitor_id"]
  x. event_id = params["event_id"]
  x.rank = params["rank"]
  x.save

  erb :"added/award_added"
end
```

Maybe this was your next step, in which case–carry on. But I wanted to make sure I pointed it out, in case it wasn't something you'd planned on doing next.

sumeetjain commented:
Other than that, however, everything looks solid. :+1:

halfghaninne commented:
Regarding your /eventdetail and /award responses, it might not be that you did anything wrong, but that some of your data got deleted or corrupted along the way. If they are listing out in your html response, I'd count that as a success!

halfghaninne commented:
Regarding your single item views, for example viewing one college:

https://github.com/jillianw/web-app__swimming-competition/blob/b8c3ce8c63e12647fc5c63a69dd4834467cc7dbb/app/controllers/view_one_controller.rb#L11-L15

Line 14's job is to point to a view file, that your controller can serve up as a response to the browser. You are pointing to a view here that doesn't exist. I think you mean to be pointing to `"view_one/view_college.erb"`

Can you tell me why you are getting the data for **all** College records in line 12, if the goal is to display information about **one**?
